pyquil/noise.py

```diff
@@ -0,0 +1,566 @@
+    :param float gate_time_1q: The duration of the one-qubit gates, namely RX(+pi/2) and RX(-pi/2).
+        By default, this is 50 ns.
+    :param float gate_time_2q: The duration of the two-qubit gates, namely CZ.
+        By default, this is 150 ns.
+    :param Union[Dict[int,float],float] ro_fidelity: The readout assignment fidelity
+        :math:`F = (p(0|0) + p(1|1))/2` either globally or in a dictionary indexed by qubit id.
+    :return: A NoiseModel with the appropriate Kraus operators defined.
+    """
+    all_qubits = set(sum(([t.index for t in g.qubits] for g in gates), []))
+    if isinstance(T1, dict):
+        all_qubits.update(T1.keys())
+    if isinstance(T2, dict):
+        all_qubits.update(T2.keys())
+    if isinstance(ro_fidelity, dict):
+        all_qubits.update(ro_fidelity.keys())
+
+    if not isinstance(T1, dict):
+        T1 = {q: T1 for q in all_qubits}
+
+    if T2 is None:
+        T2 = {q: T1 / 2. for q, T1 in T1.items()}
+    elif not isinstance(T2, dict):
+        T2 = {q: T2 for q in all_qubits}
+
+    # T2 can be at most T1/2
+    T2 = {q: min(qT2, T1.get(q, INFTY) / 2.) for q, qT2 in T2.items()}
+
+    if not isinstance(ro_fidelity, dict):
+        ro_fidelity = {q: ro_fidelity for q in all_qubits}
+
+    noisy_identities_1q = {
+        q: damping_after_dephasing(T1.get(q, INFTY), T2.get(q, INFTY), gate_time_1q)
+        for q in all_qubits
+    }
+    noisy_identities_2q = {
+        q: damping_after_dephasing(T1.get(q, INFTY), T2.get(q, INFTY), gate_time_2q)
+        for q in all_qubits
+    }
+    kraus_maps = []
+    for g in gates:
+        targets = tuple(t.index for t in g.qubits)
+        key = (g.name, tuple(g.params))
+        if g.name in NO_NOISE:
+            continue
+        if key in SINGLE_Q:
+            matrix, _ = SINGLE_Q[key]
+            noisy_I = noisy_identities_1q[targets[0]]
+        elif key in TWO_Q:
+            matrix, _ = TWO_Q[key]
+            # note this ordering of the tensor factors is necessary due to how the QVM orders
+            # the wavefunction basis
+            noisy_I = tensor_kraus_maps(noisy_identities_2q[targets[1]],
+                                        noisy_identities_2q[targets[0]])
+        else:
+            raise ValueError("Cannot create noisy version of {}. ".format(g) +
+                             "Please restrict yourself to CZ, RX(+/-pi/2), I, RZ(theta)")
+        kraus_maps.append(KrausModel(g.name, tuple(g.params), targets,
+                                     combine_kraus_maps(noisy_I, [matrix]),
+                                     # FIXME (Nik): compute actual avg gate fidelity for this simple
+                                     # noise model
+                                     1.0))
+    aprobs = {}
+    for q, f_ro in ro_fidelity.items():
+        aprobs[q] = np.array([[f_ro, 1. - f_ro],
+                              [1. - f_ro, f_ro]])
+
+    # FIXME (Nik): decide on whether isa_name is set to something more useful
```

Comment:
did you want to fix this before merge?

Reply:
I was waiting for #276 to get merged and then make the call.

Reply:
Ok, I removed it.